server/src/main/java/org/opensearch/action/search/SearchRequestOperationsListener.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+
+import java.util.List;
+
+/**
+ * A listener for search, fetch and context events at the coordinator node level
+ *
+ * @opensearch.internal
+ */
+public interface SearchRequestOperationsListener {
+
+    void onDFSPreQueryPhaseStart(SearchPhaseContext context);
+
+    void onDFSPreQueryPhaseFailure(SearchPhaseContext context);
+
+    void onDFSPreQueryPhaseEnd(SearchPhaseContext context, long tookTime);
+
+    void onCanMatchPhaseStart(SearchPhaseContext context);
+
+    void onCanMatchPhaseFailure(SearchPhaseContext context);
+
+    void onCanMatchPhaseEnd(SearchPhaseContext context, long tookTime);
+
+    void onQueryPhaseStart(SearchPhaseContext context);
+
+    void onQueryPhaseFailure(SearchPhaseContext context);
+
+    void onQueryPhaseEnd(SearchPhaseContext context, long tookTime);
+
+    void onFetchPhaseStart(SearchPhaseContext context);
+
+    void onFetchPhaseFailure(SearchPhaseContext context);
+
+    void onFetchPhaseEnd(SearchPhaseContext context, long tookTime);
+
+    void onExpandSearchPhaseStart(SearchPhaseContext context);
+
+    void onExpandSearchPhaseFailure(SearchPhaseContext context);
+
+    void onExpandSearchPhaseEnd(SearchPhaseContext context, long tookTime);
```

Comment:
Couldn't this whole thing just be `onPhase{Start, End, Failure}(SearchPhaseName, SearchPhaseContext)`?

In particular, if we're going to do generic `SearchRequestOperationsListener` interface, it's a lot more flexible if implementers don't break if we add new phases in the future. (They may get called with a new phase, but I think we can expect a listener to either handle all phases generically or only take action for certain phases, in which case they would just ignore any new/unknown phases.)

Reply:
We did evaluate to make it generic onPhaseStart, onPhaseEnd etc as part of interface. But then all the subscribers to this listener will need to have additional logic to determine the phase type(query/fetch etc) and accordingly compute stats for each. So essentially we are pushing this additional work down to subscribers which might be redundant. Instead adding it at this interface level makes it easy for subscribers to handle/compute stats for each phase type. This is how shard level operation listener also works currently (SearchOperationListener).

Plus SearchRequestStats(one of the subscriber to this listener) and other stats(to follow) are shared across requests, so having generic onPhaseEnd/onPhaseStart might put more thread contention(though not much) when compared to individual onFetchPhaseEnd/onQueryPhaseEnd etc.

Regarding adding new phases, we can always use ```default``` at interface level if we don't intend to use in all implementers. Plus I don't forsee many phases being added for this.

Let me know what you think.

Reply:
I don't think `SearchOperationListener` is a great template for this:

1. It's not just looking at search phases -- it's also tracking create/free reader context.
2. It's not looking at all search phases -- just query + fetch.
3. There wasn't an enum of available of search phases when it was created, but there is now.

In this case, since we're specifically looking at search phases (all search phases and only search phases), it makes sense to use the enumeration of all valid search phases.

> Plus SearchRequestStats(one of the subscriber to this listener) and other stats(to follow) are shared across requests, so having generic onPhaseEnd/onPhaseStart might put more thread contention(though not much) when compared to individual onFetchPhaseEnd/onQueryPhaseEnd etc.

When SearchRequestStats is initialized, it can initialize the stats elements for each phase and keep them in an immutable EnumMap that can be safely accessed without locking. Incrementing/decrementing the individual counters would have the same level of thread contention whether they're held by a map element or held directly as a field in SearchRequestStats.

Reply:
>I don't think SearchOperationListener is a great template for this:

>It's not just looking at search phases -- it's also tracking create/free reader context.

I am not sure what you exactly mean here by reader context. SearchOperationListener is only tracking inidividual phases start/end and also passes SearchPhaseContext which has additional data around the task/request which can be used later if needed.

> It's not looking at all search phases -- just query + fetch.

Again not very clear to me, but we are tracking canMatch, expand and dfs phase as well apart from above 2.

> There wasn't an enum of available of search phases when it was created, but there is now.

I don't think this is really a problem or something which made us implement it like the way we did. 


**So you are suggesting something like(so that I can understand better):**

```
interface SearchRequestOperationsListener {
    void onPhaseStart(SearchPhase phase);
    void onPhaseEnd(SearchPhase phase);
   // ....
}

// one of the subscribers to above listener
SearchRequestStats implements SearchRequestOperationsListener {
      void onPhaseStart(SearchPhase phase) {
             // write logic to determine phase by enum and accordingly update counter.
            // As u mentioned something like keeping a EnumMap or similar.
       }
       
      void onPhaseEnd(SearchPhase phase, long tookTime) {
           // same logic as above to update counters etc.
      } 
}

// if we have another subscriber, we need to repeat same logic as above. 
```

While my logic is segregating phase type on listener interface itself.
```
interface SearchRequestOperationsListener {
    void onQueryPhaseStart(SearchPhase phase);
}

SearchRequestStats implements SearchRequestOperationsListener {
     void onQueryPhaseStart(SearchPhase phase) {
       // current logic to compute stats
     }
}

```

Again IMO second(current) approach might be more verbose but is simple as well(can be subjective). The concern you had around breaking changes when we add new phases can be easily avoided by using default method in interfaces plus would be rare to introduce new phases often.

Reply:
This interface feel too verbose and I kind of echo to what @msfroh is saying in his initial comments. If we go with the first approach and use the `SearchPhaseName` we can easily record the stats. This will also ensure that going forward if there are new phases getting added we the stats code will work without any change.

Reply:
This interface feel too verbose and I kind of echo to what @msfroh is saying in his initial comments. If we go with the first approach and use the SearchPhaseName we can easily record the stats. This will also ensure that going forward if there are new phases getting added we the stats code will work without any change.

Reply:
Okay. I will try to change it then if both of you are inclined towards that.

Reply:
@msfroh @navneet1v I have made changes to the interface as suggested, can you review it again please?